File: sdk/go/tonebridge/websocket.go
package tonebridge

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

const (
	// WebSocket message types
	MessageTypeTransform     = "transform"
	MessageTypeAnalyze       = "analyze"
	MessageTypeAutoTransform = "auto_transform"
	MessageTypeSubscribe     = "subscribe"
	MessageTypeUnsubscribe   = "unsubscribe"
	MessageTypePing          = "ping"
	MessageTypePong          = "pong"
	MessageTypeError         = "error"
	MessageTypeEvent         = "event"

	// Default WebSocket settings
	defaultPingInterval   = 30 * time.Second
	defaultPongTimeout    = 10 * time.Second
	defaultReconnectDelay = 5 * time.Second
	defaultMaxReconnects  = 10
	writeTimeout          = 10 * time.Second
)

// WebSocketClient handles WebSocket connections
type WebSocketClient struct {
	url            string
	token          string
	conn           *websocket.Conn
	mu             sync.RWMutex
	connected      bool
	reconnecting   bool
	reconnectCount int
	maxReconnects  int
	reconnectDelay time.Duration

	// Channels
	send      chan []byte
	receive   chan *WebSocketMessage
	done      chan struct{}
	reconnect chan struct{}

	// Callbacks
	onConnect    func()
	onDisconnect func()
	onMessage    func(interface{})
	onError      func(error)

	// Subscriptions
	subscriptions map[string]bool
	subMu         sync.RWMutex

	// Context for cancellation
	ctx    context.Context
	cancel context.CancelFunc
}

// WebSocketOption is a function that configures the WebSocket client
type WebSocketOption func(*WebSocketClient)

// NewWebSocketClient creates a new WebSocket client
func NewWebSocketClient(url, token string, opts ...WebSocketOption) *WebSocketClient {
	ctx, cancel := context.WithCancel(context.Background())

	client := &WebSocketClient{
		url:            url,
		token:          token,
		maxReconnects:  defaultMaxReconnects,
		reconnectDelay: defaultReconnectDelay,
		send:           make(chan []byte, 256),
		receive:        make(chan *WebSocketMessage, 256),
		done:           make(chan struct{}),
		reconnect:      make(chan struct{}, 1),
		subscriptions:  make(map[string]bool),
		ctx:            ctx,
		cancel:         cancel,
	}

	// Apply options
	for _, opt := range opts {
		opt(client)
	}

	return client
}

// WithWSCallbacks sets WebSocket callbacks
func WithWSCallbacks(onConnect, onDisconnect func(), onMessage func(interface{}), onError func(error)) WebSocketOption {
	return func(c *WebSocketClient) {
		c.onConnect = onConnect
		c.onDisconnect = onDisconnect
		c.onMessage = onMessage
		c.onError = onError
	}
}

// WithMaxReconnects sets the maximum reconnect attempts
func WithMaxReconnects(max int) WebSocketOption {
	return func(c *WebSocketClient) {
		c.maxReconnects = max
	}
}

// WithReconnectDelay sets the reconnect delay
func WithReconnectDelay(delay time.Duration) WebSocketOption {
	return func(c *WebSocketClient) {
		c.reconnectDelay = delay
	}
}

// Connect establishes WebSocket connection
func (c *WebSocketClient) Connect() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.connected {
		return nil
	}

	headers := http.Header{}
	if c.token != "" {
		headers.Add("Authorization", "Bearer "+c.token)
	}

	dialer := websocket.Dialer{
		HandshakeTimeout: 10 * time.Second,
	}

	conn, _, err := dialer.Dial(c.url, headers)
	if err != nil {
		c.handleError(fmt.Errorf("websocket dial error: %w", err))
		return err
	}

	c.conn = conn
	c.connected = true
	c.reconnectCount = 0

	// Set up ping/pong handlers
	c.conn.SetReadDeadline(time.Now().Add(defaultPongTimeout + defaultPingInterval))
	c.conn.SetPongHandler(func(string) error {
		c.conn.SetReadDeadline(time.Now().Add(defaultPongTimeout + defaultPingInterval))
		return nil
	})

	// Start goroutines
	go c.readPump()
	go c.writePump()
	go c.pingLoop()

	// Resubscribe to previous subscriptions
	c.resubscribe()

	// Call onConnect callback
	if c.onConnect != nil {
		c.onConnect()
	}

	return nil
}

// Disconnect closes the WebSocket connection
func (c *WebSocketClient) Disconnect() {
	c.mu.Lock()
	defer c.mu.Unlock()

	if !c.connected {
		return
	}

	c.connected = false
	c.cancel()
	close(c.done)

	if c.conn != nil {
		c.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
		c.conn.Close()
	}

	// Call onDisconnect callback
	if c.onDisconnect != nil {
		c.onDisconnect()
	}
}

// IsConnected returns connection status
func (c *WebSocketClient) IsConnected() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.connected
}

// Send sends a message through WebSocket
func (c *WebSocketClient) Send(message interface{}) error {
	if !c.IsConnected() {
		return ErrNoConnection
	}

	data, err := json.Marshal(message)
	if err != nil {
		return fmt.Errorf("failed to marshal message: %w", err)
	}

	select {
	case c.send <- data:
		return nil
	case <-time.After(writeTimeout):
		return fmt.Errorf("send timeout")
	}
}

// Subscribe subscribes to a channel or event
func (c *WebSocketClient) Subscribe(channel string) error {
	c.subMu.Lock()
	c.subscriptions[channel] = true
	c.subMu.Unlock()

	msg := map[string]interface{}{
		"type":    MessageTypeSubscribe,
		"channel": channel,
	}

	return c.Send(msg)
}

// Unsubscribe unsubscribes from a channel or event
func (c *WebSocketClient) Unsubscribe(channel string) error {
	c.subMu.Lock()
	delete(c.subscriptions, channel)
	c.subMu.Unlock()

	msg := map[string]interface{}{
		"type":    MessageTypeUnsubscribe,
		"channel": channel,
	}

	return c.Send(msg)
}

// TransformRealtime performs real-time transformation
func (c *WebSocketClient) TransformRealtime(req *TransformRequest, callback func(*TransformResponse, error)) error {
	if !c.IsConnected() {
		return ErrNoConnection
	}

	msgID := generateMessageID()
	msg := map[string]interface{}{
		"type": MessageTypeTransform,
		"id":   msgID,
		"data": req,
	}

	// Register callback for response
	c.registerCallback(msgID, func(data interface{}) {
		resp, ok := data.(*TransformResponse)
		if !ok {
			callback(nil, fmt.Errorf("invalid response type"))
			return
		}
		callback(resp, nil)
	})

	return c.Send(msg)
}

// AnalyzeRealtime performs real-time analysis
func (c *WebSocketClient) AnalyzeRealtime(req *AnalyzeRequest, callback func(*AnalyzeResponse, error)) error {
	if !c.IsConnected() {
		return ErrNoConnection
	}

	msgID := generateMessageID()
	msg := map[string]interface{}{
		"type": MessageTypeAnalyze,
		"id":   msgID,
		"data": req,
	}

	// Register callback for response
	c.registerCallback(msgID, func(data interface{}) {
		resp, ok := data.(*AnalyzeResponse)
		if !ok {
			callback(nil, fmt.Errorf("invalid response type"))
			return
		}
		callback(resp, nil)
	})

	return c.Send(msg)
}

// AutoTransformRealtime performs real-time auto-transformation
func (c *WebSocketClient) AutoTransformRealtime(ctx *MessageContext, callback func(*AutoTransformResponse, error)) error {
	if !c.IsConnected() {
		return ErrNoConnection
	}

	msgID := generateMessageID()
	msg := map[string]interface{}{
		"type": MessageTypeAutoTransform,
		"id":   msgID,
		"data": ctx,
	}

	// Register callback for response
	c.registerCallback(msgID, func(data interface{}) {
		resp, ok := data.(*AutoTransformResponse)
		if !ok {
			callback(nil, fmt.Errorf("invalid response type"))
			return
		}
		callback(resp, nil)
	})

	return c.Send(msg)
}

// readPump reads messages from WebSocket
func (c *WebSocketClient) readPump() {
	defer func() {
		c.mu.Lock()
		c.connected = false
		c.mu.Unlock()
		c.conn.Close()
		c.tryReconnect()
	}()

	for {
		select {
		case <-c.done:
			return
		default:
			var msg WebSocketMessage
			err := c.conn.ReadJSON(&msg)
			if err != nil {
				if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
					c.handleError(fmt.Errorf("websocket read error: %w", err))
				}
				return
			}

			// Handle message
			c.handleMessage(&msg)
		}
	}
}

// writePump writes messages to WebSocket
func (c *WebSocketClient) writePump() {
	ticker := time.NewTicker(defaultPingInterval)
	defer func() {
		ticker.Stop()
		c.conn.Close()
	}()

	for {
		select {
		case message, ok := <-c.send:
			c.conn.SetWriteDeadline(time.Now().Add(writeTimeout))
			if !ok {
				c.conn.WriteMessage(websocket.CloseMessage, []byte{})
				return
			}

			if err := c.conn.WriteMessage(websocket.TextMessage, message); err != nil {
				c.handleError(fmt.Errorf("websocket write error: %w", err))
				return
			}

		case <-ticker.C:
			c.conn.SetWriteDeadline(time.Now().Add(writeTimeout))
			if err := c.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
				return
			}

		case <-c.done:
			return
		}
	}
}

// pingLoop sends periodic ping messages
func (c *WebSocketClient) pingLoop() {
	ticker := time.NewTicker(defaultPingInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			msg := map[string]interface{}{
				"type": MessageTypePing,
			}
			if err := c.Send(msg); err != nil {
				c.handleError(fmt.Errorf("ping error: %w", err))
			}

		case <-c.done:
			return
		}
	}
}

// handleMessage processes incoming messages
func (c *WebSocketClient) handleMessage(msg *WebSocketMessage) {
	switch msg.Type {
	case MessageTypePong:
		// Pong received, connection is alive

	case MessageTypeError:
		if c.onError != nil {
			c.onError(fmt.Errorf("server error: %v", msg.Data))
		}

	case MessageTypeEvent:
		if c.onMessage != nil {
			c.onMessage(msg.Data)
		}

	default:
		// Check for registered callbacks
		if callback := c.getCallback(msg.ID); callback != nil {
			callback(msg.Data)
			c.removeCallback(msg.ID)
		} else if c.onMessage != nil {
			c.onMessage(msg)
		}
	}
}

// handleError handles errors
func (c *WebSocketClient) handleError(err error) {
	if c.onError != nil {
		c.onError(err)
	}
}

// tryReconnect attempts to reconnect
func (c *WebSocketClient) tryReconnect() {
	c.mu.Lock()
	if c.reconnecting || c.reconnectCount >= c.maxReconnects {
		c.mu.Unlock()
		return
	}
	c.reconnecting = true
	c.reconnectCount++
	c.mu.Unlock()

	time.Sleep(c.reconnectDelay)

	if err := c.Connect(); err != nil {
		c.mu.Lock()
		c.reconnecting = false
		c.mu.Unlock()
		c.tryReconnect()
	} else {
		c.mu.Lock()
		c.reconnecting = false
		c.mu.Unlock()
	}
}

// resubscribe resubscribes to previous subscriptions
func (c *WebSocketClient) resubscribe() {
	c.subMu.RLock()
	channels := make([]string, 0, len(c.subscriptions))
	for channel := range c.subscriptions {
		channels = append(channels, channel)
	}
	c.subMu.RUnlock()

	for _, channel := range channels {
		msg := map[string]interface{}{
			"type":    MessageTypeSubscribe,
			"channel": channel,
		}
		c.Send(msg)
	}
}

// Callback management
var (
	callbacks   = make(map[string]func(interface{}))
	callbacksMu sync.RWMutex
)

func registerCallback(id string, callback func(interface{})) {
	callbacksMu.Lock()
	defer callbacksMu.Unlock()
	callbacks[id] = callback
}

func (c *WebSocketClient) registerCallback(id string, callback func(interface{})) {
	registerCallback(id, callback)
}

func getCallback(id string) func(interface{}) {
	callbacksMu.RLock()
	defer callbacksMu.RUnlock()
	return callbacks[id]
}

func (c *WebSocketClient) getCallback(id string) func(interface{}) {
	return getCallback(id)
}

func removeCallback(id string) {
	callbacksMu.Lock()
	defer callbacksMu.Unlock()
	delete(callbacks, id)
}

func (c *WebSocketClient) removeCallback(id string) {
	removeCallback(id)
}

// generateMessageID generates a unique message ID
func generateMessageID() string {
	return fmt.Sprintf("%d", time.Now().UnixNano())
}

File: sdk/python/tonebridge/websocket.py
"""
WebSocket Client for real-time communication
"""

import json
import threading
import time
from typing import Optional, Callable, Any, Dict
import websocket
from .exceptions import WebSocketError


class WebSocketClient:
    """WebSocket client for real-time ToneBridge communication"""
    
    def __init__(
        self,
        url: str,
        token: Optional[str] = None,
        reconnect: bool = True,
        max_reconnect_attempts: int = 5,
        reconnect_interval: int = 5,
        on_connect: Optional[Callable] = None,
        on_disconnect: Optional[Callable] = None,
        on_message: Optional[Callable] = None,
        on_error: Optional[Callable] = None,
    ):
        """
        Initialize WebSocket client
        
        Args:
            url: WebSocket URL
            token: Authentication token
            reconnect: Enable auto-reconnect
            max_reconnect_attempts: Maximum reconnection attempts
            reconnect_interval: Interval between reconnection attempts (seconds)
            on_connect: Connection callback
            on_disconnect: Disconnection callback
            on_message: Message callback
            on_error: Error callback
        """
        self.url = url
        self.token = token
        self.reconnect = reconnect
        self.max_reconnect_attempts = max_reconnect_attempts
        self.reconnect_interval = reconnect_interval
        
        # Callbacks
        self.on_connect = on_connect
        self.on_disconnect = on_disconnect
        self.on_message = on_message
        self.on_error = on_error
        
        # State
        self.ws: Optional[websocket.WebSocketApp] = None
        self.is_connected = False
        self.reconnect_attempts = 0
        self.message_queue = []
        self.thread: Optional[threading.Thread] = None
        self.stop_flag = threading.Event()
    
    def connect(self) -> None:
        """Connect to WebSocket server"""
        if self.is_connected:
            return
        
        # Add token to URL if provided
        ws_url = self.url
        if self.token:
            separator = "&" if "?" in ws_url else "?"
            ws_url = f"{ws_url}{separator}token={self.token}"
        
        self.ws = websocket.WebSocketApp(
            ws_url,
            on_open=self._on_open,
            on_message=self._on_message,
            on_error=self._on_error,
            on_close=self._on_close,
        )
        
        # Start WebSocket in separate thread
        self.stop_flag.clear()
        self.thread = threading.Thread(target=self._run_forever)
        self.thread.daemon = True
        self.thread.start()
    
    def disconnect(self) -> None:
        """Disconnect from WebSocket server"""
        self.stop_flag.set()
        self.is_connected = False
        
        if self.ws:
            self.ws.close()
            self.ws = None
        
        if self.thread and self.thread.is_alive():
            self.thread.join(timeout=5)
    
    def send(self, message_type: str, data: Any) -> None:
        """
        Send message through WebSocket
        
        Args:
            message_type: Type of message
            data: Message data
        """
        message = {
            "type": message_type,
            "data": data,
            "timestamp": time.time(),
            "id": self._generate_id(),
        }
        
        if self.is_connected and self.ws:
            try:
                self.ws.send(json.dumps(message))
            except Exception as e:
                self._handle_error(e)
        else:
            # Queue message if not connected
            self.message_queue.append(message)
    
    def send_transform(self, data: Dict[str, Any]) -> None:
        """
        Send transform request
        
        Args:
            data: Transform data
        """
        self.send("transform", data)
    
    def send_analyze(self, data: Dict[str, Any]) -> None:
        """
        Send analyze request
        
        Args:
            data: Analyze data
        """
        self.send("analyze", data)
    
    def is_connected(self) -> bool:
        """Check if connected to server"""
        return self.is_connected
    
    def _run_forever(self) -> None:
        """Run WebSocket connection loop"""
        while not self.stop_flag.is_set():
            try:
                self.ws.run_forever()
                
                if not self.stop_flag.is_set() and self.reconnect:
                    self._handle_reconnect()
                else:
                    break
            except Exception as e:
                self._handle_error(e)
                
                if self.reconnect and self.reconnect_attempts < self.max_reconnect_attempts:
                    time.sleep(self.reconnect_interval)
                else:
                    break
    
    def _on_open(self, ws) -> None:
        """Handle connection open"""
        self.is_connected = True
        self.reconnect_attempts = 0
        
        # Send queued messages
        self._flush_message_queue()
        
        if self.on_connect:
            try:
                self.on_connect()
            except Exception as e:
                self._handle_error(e)
    
    def _on_message(self, ws, message) -> None:
        """Handle incoming message"""
        try:
            data = json.loads(message)
            
            if self.on_message:
                self.on_message(data)
        except json.JSONDecodeError as e:
            self._handle_error(WebSocketError(f"Failed to parse message: {e}"))
        except Exception as e:
            self._handle_error(e)
    
    def _on_error(self, ws, error) -> None:
        """Handle WebSocket error"""
        self._handle_error(error)
    
    def _on_close(self, ws, close_status_code, close_msg) -> None:
        """Handle connection close"""
        self.is_connected = False
        
        if self.on_disconnect:
            try:
                self.on_disconnect()
            except Exception as e:
                self._handle_error(e)
    
    def _handle_error(self, error: Exception) -> None:
        """Handle error"""
        if self.on_error:
            try:
                self.on_error(error)
            except Exception:
                pass  # Ignore errors in error handler
    
    def _handle_reconnect(self) -> None:
        """Handle reconnection"""
        if self.reconnect_attempts >= self.max_reconnect_attempts:
            return
        
        self.reconnect_attempts += 1
        time.sleep(self.reconnect_interval * self.reconnect_attempts)
    
    def _flush_message_queue(self) -> None:
        """Send queued messages"""
        while self.message_queue and self.is_connected:
            message = self.message_queue.pop(0)
            try:
                if self.ws:
                    self.ws.send(json.dumps(message))
            except Exception as e:
                # Re-queue message on failure
                self.message_queue.insert(0, message)
                self._handle_error(e)
                break
    
    def _generate_id(self) -> str:
        """Generate unique message ID"""
        import uuid
        return str(uuid.uuid4())